Sources/Purchasing/Purchases/TransactionPoster.swift

```diff
@@ -0,0 +1,306 @@
+    )
+
+    func markSyncedIfNeeded(
+        subscriberAttributes: SubscriberAttribute.Dictionary?,
+        error: BackendError?
+    )
+
+}
+
+// swiftlint:disable function_parameter_count
+
+final class TransactionPoster: TransactionPosterType {
+
+    private let productsManager: ProductsManagerType
+    private let receiptFetcher: ReceiptFetcher
+    private let currentUserProvider: CurrentUserProvider
+    private let attribution: Attribution
+    private let backend: Backend
+    private let paymentQueueWrapper: EitherPaymentQueueWrapper
+    private let systemInfo: SystemInfo
+    private let operationDispatcher: OperationDispatcher
+
+    init(
+        productsManager: ProductsManagerType,
+        receiptFetcher: ReceiptFetcher,
+        currentUserProvider: CurrentUserProvider,
+        attribution: Attribution,
+        backend: Backend,
+        paymentQueueWrapper: EitherPaymentQueueWrapper,
+        systemInfo: SystemInfo,
+        operationDispatcher: OperationDispatcher
+    ) {
+        self.productsManager = productsManager
+        self.receiptFetcher = receiptFetcher
+        self.currentUserProvider = currentUserProvider
+        self.attribution = attribution
+        self.backend = backend
+        self.paymentQueueWrapper = paymentQueueWrapper
+        self.systemInfo = systemInfo
+        self.operationDispatcher = operationDispatcher
+    }
+
+    func handlePurchasedTransaction(_ transaction: StoreTransaction,
+                                    presentedOfferingID: String?,
+                                    storefront: StorefrontType?,
+                                    source: PurchaseSource,
+                                    completion: @escaping PurchaseCompletedBlock) {
+        self.receiptFetcher.receiptData(
+            refreshPolicy: self.refreshRequestPolicy(forProductIdentifier: transaction.productIdentifier)
+        ) { receiptData in
+            if let receiptData = receiptData, !receiptData.isEmpty {
+                self.fetchProductsAndPostReceipt(
+                    withTransaction: transaction,
+                    receiptData: receiptData,
+                    source: source,
+                    storefront: storefront,
+                    presentedOfferingID: presentedOfferingID,
+                    completion: completion
+                )
+            } else {
+                self.handleReceiptPost(withTransaction: transaction,
+                                       result: .failure(.missingReceiptFile()),
+                                       subscriberAttributes: nil,
+                                       completion: completion)
+            }
+        }
+    }
+
+    func finishTransactionIfNeeded(
+        _ transaction: StoreTransactionType,
+        completion: @escaping @Sendable @MainActor () -> Void
+    ) {
+        @Sendable
+        func complete() {
+            self.operationDispatcher.dispatchOnMainActor(completion)
+        }
```

Comment:
I realize this was already there, but I'm not sure it's necessarily easier to read having this one-line function called twice vs just calling the same line twice manually

Reply:
lmk if it was done for `@Sendable` purposes

Reply:
I don't think so, it was to remove the duplication so we didn't accidentally use `dispatchOnMainQueue` or something in one of them.